app/models/conversion_host.rb

```diff
@@ -9,28 +9,109 @@ class ConversionHost < ApplicationRecord
 
   # To be eligible, a conversion host must have the following properties
   #  - A transport mechanism is configured for source (set by 3rd party)
-  #  - Credentials are set on the resource
+  #  - Credentials are set on the resource and SSH connection works
   #  - The number of concurrent tasks has not reached the limit
   def eligible?
-    source_transport_method.present? && check_resource_credentials && check_concurrent_tasks
+    source_transport_method.present? && check_ssh_connection && check_concurrent_tasks
   end
 
   def check_concurrent_tasks
     max_tasks = max_concurrent_tasks || Settings.transformation.limits.max_concurrent_tasks_per_host
     active_tasks.size < max_tasks
   end
 
-  def check_resource_credentials
-    send("check_resource_credentials_#{resource.ext_management_system.emstype}")
+  def check_ssh_connection
+    connect_ssh { |ssu| ssu.shell_exec('uname -a') }
+    true
+  rescue => e
+    false
   end
 
   def source_transport_method
     return 'vddk' if vddk_transport_supported
     return 'ssh' if ssh_transport_supported
   end
 
+  def ipaddress(family = nil)
+    ips = ipaddresses
+    return ips.first unless %w(ipv4 ipv6).include?(family)
```

Comment:
The idea was to filter out invalid family.

Reply:
You could always `raise unless %w(ipv4 ipv6).include?(family)` if you're worried about it, but if someone passes in an invalid family then `ipaddresses.detect { |ip| IPAddr.new(ip).send("#{family}?") } will do that for you if someone from the future tries `ip.ipv9?`

Reply:
But I don't want to raise, rather gracefully fallback to first IP address whatever its family.
What's the common approach in ManageIQ code ? raising or falling back ?

Reply:
Falling back to the first IP address when you asked for an invalid family is confusing, makes it seem like it succeeded.
We usually prefer to blow up on invalid input rather than hide potential bugs that could creep up later on.